### PR TITLE
feat/opencl: getContextInfo implementation

### PR DESCRIPTION
**Pull request by drahnr**
* [x] compiles
* [x] Implement tests
* [x] investigate: `unused_qualifications` was removed, which I am not sure what it does or why it needed to be there in the first place, adding it causes `std::mem::size_of` to fail
* [x] meaningful docstrings

### CODE REVIEW

**Comment by drahnr** [comment]
+r

File: tests/framework_opencl_specs.rs [hunk 1]
```diff
@@ -35,4 +36,14 @@ mod framework_opencl_spec {
         let ctx = frm.new_device(&frm.hardwares()[0..1]).unwrap();
         Queue::new(&ctx, &frm.hardwares()[0..1][0], None).unwrap();
     }
+
+    #[test]
+    fn it_queries_context_info() {
+        let frm = OpenCL::new();
+        let ctx = frm.new_device(&frm.hardwares()[0..1]).unwrap();
+        println!("ReferenceCount: {:?}", ctx.get_context_info(ContextInfoQuery::ReferenceCount));
+        println!("NumDevices: {:?}", ctx.get_context_info(ContextInfoQuery::NumDevices));
```
Comment (ghost) [inline tests/framework_opencl_specs.rs:45]: given that we ought to be able to expect that the compiler has *something* that will respond to openCL queries, how about some `is_ok()` as a basic test?

```
assert!(ctx.get_context_info(ContextInfoQuery::ReferenceCount).is_ok());        
assert!(ctx.get_context_info(ContextInfoQuery::NumDevices).is_ok());
assert!(ctx.get_context_info(ContextInfoQuery::Devices).is_ok());
assert!(ctx.get_context_info(ContextInfoQuery::Properties).is_ok());
```

**Review by ghost — COMMENTED** [review]
The reference and device counts look strange on my machine, incidentally

**Comment by ghost** [comment]
Here is example output from running the test on my machine:

```
ReferenceCount: Ok(ReferenceCount(3252990176))
NumDevices: Ok(NumDevices(3252990176))
Devices: Ok(Devices([Device { id: 3253164144, name: None, device_type: None, compute_units: None, version: None, vendor: None }, Device { id: 32690, name: None, device_type: None, compute_units: None, version: None, vendor: None }]))
0
Properties: Ok(Properties([]))
```

NumDevices and ReferenceCount look off.

**Comment by subversive-owl** [comment]
There are a lot of casts one after the other in the `CL_CONTEXT_DEVICES` match arm, but only one in `CL_CONTEXT_REFERENCE_COUNT`, so casting itself might not be the issue...

**Comment by subversive-owl** [comment]
Here is another test output:

```
running 1 test
ReferenceCount: Ok(ReferenceCount(1613017312))
NumDevices: Ok(NumDevices(1613017312))
Devices: Ok(Devices([Device { id: 1613191280, name: None, device_type: None, compute_units: None, version: None, vendor: None }, Device { id: 32558, name: None, device_type: None, compute_units: None, version: None, vendor: None }]))
0
Properties: Ok(Properties([]))
test framework_opencl_spec::it_queries_context_info ... ok
```

**Comment by drahnr** [comment]
Can you run something like `clinfo` and attach the the output?

That does look off...

**Comment by subversive-owl** [comment]
This looks better on the output, although the second device is strange:

```
running 1 test
ReferenceCount: Ok(ReferenceCount(1))
NumDevices: Ok(NumDevices(1))
Devices: Ok(Devices([Device { id: 3458685040, name: None, device_type: None, compute_units: None, version: None, vendor: None }, Device { id: 32547, name: None, device_type: None, compute_units: None, version: None, vendor: None }]))
0
Properties: Ok(Properties([]))
test framework_opencl_spec::it_queries_context_info ... ok
```

clinfo:

```
Number of platforms                               1
  Platform Name                                   NVIDIA CUDA
  Platform Vendor                                 NVIDIA Corporation
  Platform Version                                OpenCL 1.2 CUDA 8.0.0
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_fp64 cl_khr_byte_addressable_store cl_khr_icd cl_khr_gl_sharing cl_nv_compiler_options cl_nv_device_attribute_query cl_nv_pragma_unroll cl_nv_copy_opts cl_nv_create_buffer
  Platform Extensions function suffix             NV

  Platform Name                                   NVIDIA CUDA
Number of devices                                 1
  Device Name                                     GeForce GTX 1060 6GB
  Device Vendor                                   NVIDIA Corporation
  Device Vendor ID                                0x10de
  Device Version                                  OpenCL 1.2 CUDA
  Driver Version                                  375.66
  Device OpenCL C Version                         OpenCL C 1.2 
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Device Topology (NV)                            PCI-E, 01:00.0
  Max compute units                               10
  Max clock frequency                             1771MHz
  Compute Capability (NV)                         6.1
  Device Partition                                (core)
    Max number of sub-devices                     1
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             1024x1024x64
  Max work group size                             1024
  Preferred work group size multiple              32
  Warp size (NV)                                  32
  Preferred / native vector sizes                 
    char                                                 1 / 1       
    short                                                1 / 1       
    int                                                  1 / 1       
    long                                                 1 / 1       
    half                                                 0 / 0        (n/a)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Address bits                                    64, Little-Endian
  Global memory size                              6359285760 (5.923GiB)
  Error Correction support                        No
  Max memory allocation                           1589821440 (1.481GiB)
  Unified memory for Host and Device              No
  Integrated memory (NV)                          No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       4096 bits (512 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        163840
  Global Memory cache line                        128 bytes
  Image support                                   Yes
    Max number of samplers per kernel             32
    Max size for 1D images from buffer            134217728 pixels
    Max 1D or 2D image array size                 2048 images
    Max 2D image size                             16384x32768 pixels
    Max 3D image size                             16384x16384x16384 pixels
    Max number of read image args                 256
    Max number of write image args                16
  Local memory type                               Local
  Local memory size                               49152 (48KiB)
  Registers per block (NV)                        65536
  Max constant buffer size                        65536 (64KiB)
  Max number of constant args                     9
  Max size of kernel argument                     4352 (4.25KiB)
  Queue properties                                
    Out-of-order execution                        Yes
    Profiling                                     Yes
  Prefer user sync for interop                    No
  Profiling timer resolution                      1000ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
    Kernel execution timeout (NV)                 Yes
  Concurrent copy and kernel execution (NV)       Yes
    Number of async copy engines                  2
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels                                
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Device Extensions                               cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_fp64 cl_khr_byte_addressable_store cl_khr_icd cl_khr_gl_sharing cl_nv_compiler_options cl_nv_device_attribute_query cl_nv_pragma_unroll cl_nv_copy_opts cl_nv_create_buffer

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  No platform
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   No platform
  clCreateContext(NULL, ...) [default]            No platform
  clCreateContext(NULL, ...) [other]              Success [NV]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  No platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  No platform
```

**Comment by drahnr** [comment]
True... I will look into this tomorrow.

**Comment by subversive-owl** [comment]
Looks good now:

```
running 1 test
ReferenceCount: Ok(ReferenceCount(1))
NumDevices: Ok(NumDevices(1))
Devices: Ok(Devices([Device { id: 139748841052272, name: None, device_type: None, compute_units: None, version: None, vendor: None }]))
0
Properties: Ok(Properties([]))
test framework_opencl_spec::it_queries_context_info ... ok
```

I'm still not sure where that stray `0` is coming from still, but it works!

**Comment by drahnr** [comment]
Fixed that too.

**Pull request merged**